File: src/crwN2ll.f90
SUBROUTINE crwn2ll(tau2y, tau2x, sig2, b, bd, sig2d, delta, x, y, loctype, stay, &
                    ay, ax, Py, Px, lonadj, N, lly, llx)
  INTEGER N
  INTEGER loctype(N), stay(N)
  DOUBLE PRECISION lly, llx, tau2y(N), tau2x(N)
  DOUBLE PRECISION sig2(N), b(N), bd, sig2d, delta(N), x(N), y(N), lonadj(N)
  DOUBLE PRECISION ay(2,1), ax(2,1), Py(2,2), Px(2,2)
  DOUBLE PRECISION vy, vx, Fy, Fx
  DOUBLE PRECISION Z(1,2)
  DOUBLE PRECISION Ky(2,1), Kx(2,1)
  DOUBLE PRECISION Qy(2,2), Qx(2,2), T(2,2), Lx(2,2), Ly(2,2)

  !! INITIAL VALUES !!
  Z = RESHAPE((/1.0, 0.0/),(/1, 2/))
  T = 0.0
  T(1,1) = 1.0
  Qy = 0.0
  Qx = 0.0

 !! BEGIN FILTER LOOP !!
  DO i=1,N

   !! GENERATE Q AND T MATRICES !!
    IF(stay(i)==1) THEN
      Qy = 0.0
      Qx = 0.0
      T(1,2) = 0.0
      T(2,2) = 0.0
    ELSE
      Qy(1,1) = (sig2(i)/(b(i)*b(i)))*(delta(i) - 2*(1-exp(-b(i)*delta(i)))/b(i) + (1-exp(-2*b(i)*delta(i)))/(2*b(i)))
      Qy(2,1) = (sig2(i)/(b(i)*b(i)))*(1-2*exp(-b(i)*delta(i))+exp(-2*b(i)*delta(i)))/2
      Qy(1,2) = Qy(2,1)
      Qy(2,2) = (sig2(i)/b(i))*(1-exp(-2*b(i)*delta(i)))/2
      Qx = Qy/(lonadj(i)*lonadj(i))
      T(1,2) = (1-exp(-b(i)*delta(i)))/b(i)
      T(2,2) = exp(-b(i)*delta(i))
    END IF

   !! GENERAL KF LOOP !!
    IF(loctype(i)==1) THEN
      ay = MATMUL(T,ay)
      ax = MATMUL(T,ax)
      Py = MATMUL(MATMUL(T,Py),TRANSPOSE(T)) + Qy
      Px = MATMUL(MATMUL(T,Px),TRANSPOSE(T)) + Qx
    ELSE
      vy = y(i)-ay(1,1)
      vx = x(i)-ax(1,1)
      Fy = Py(1,1) + tau2y(i)
      Fx = Px(1,1) + tau2x(i)/(lonadj(i)*lonadj(i))
      Ky = MATMUL(MATMUL(T,Py),TRANSPOSE(Z))/Fy
      Kx = MATMUL(MATMUL(T,Px),TRANSPOSE(Z))/Fx
      Ly = T - MATMUL(Ky,Z)
      Lx = T - MATMUL(Kx,Z)
      ay = MATMUL(T,ay) + Ky*vy
      ax = MATMUL(T,ax) + Kx*vx
      Py = MATMUL(MATMUL(T,Py),TRANSPOSE(Ly)) + Qy
      Px = MATMUL(MATMUL(T,Px),TRANSPOSE(Lx)) + Qx
      lly = lly - (log(Fy) + vy*vy/Fy)/2
      llx = llx - (log(Fx) + vx*vx/Fx)/2
    END IF
  END DO
END SUBROUTINE crwn2ll

File: R/crwN2ll.R
"crwN2ll" <- function(theta, fixPar, y, x, loctype, delta, a1.y, a1.x,
                      P1.x, P1.y, lonAdj, mov.mf, err.mfX, err.mfY, stop.mf,
                      n.errX, n.errY, n.mov, stopMod, driftMod)
{
    N <- length(y)
    fixPar[is.na(fixPar)] <- theta
    if (!is.null(err.mfX)) {
        theta.errX <- fixPar[1:n.errX]
        tau2x <- exp(2 * err.mfX %*% theta.errX)
    } else tau2x <- 0.0
    if (!is.null(err.mfY)) {
        theta.errY <- fixPar[(n.errX + 1):(n.errX + n.errY)]
        tau2y <- exp(2 * err.mfY %*% theta.errY)
    } else tau2y <- tau2x
    theta.mov <- fixPar[(n.errX + n.errY + 1):(n.errX + n.errY + 2 * n.mov)]
    sig2 <- exp(2 * (mov.mf %*% theta.mov[1:n.mov]))
    b <- exp(2 * (mov.mf %*% theta.mov[(n.mov + 1):(2 * n.mov)]))
    stay <- rep(0, N)
    if (stopMod) {
        theta.stop <- fixPar[(n.errX + n.errY + 2 * n.mov + 1)]
        b[stop.mf != 1] <- b[stop.mf != 1] / ((1 - stop.mf[stop.mf != 1]) ^
                               exp(theta.stop))
        stay[stop.mf == 1] <- 1
    }
    if (driftMod) {
        theta.drift <- fixPar[(n.errX + n.errY + 2 * n.mov + 1):
                              (n.errX + n.errY + 2 * n.mov + 2)]
        b.drift <- b / exp(theta.drift[2])
        sig2.drift <- exp(2 * theta.drift[1])
        call.lik <- "crwdriftn2ll"
    } else {
        b.drift <- sig2.drift <- 0.0
        call.lik <- "crwn2ll"
    }
    out <- .Fortran(call.lik,
                    tau2y=as.double(tau2y),
                    tau2x=as.double(tau2x),
                    sig2=as.double(sig2),
                    b=as.double(b),
                    bd=as.double(b.drift),
                    sig2d=as.double(sig2.drift),
                    delta=as.double(delta),
                    x=as.double(x),
                    y=as.double(y),
                    loctype=as.integer(loctype),
                    stay=as.integer(stay),
                    ay=as.double(a1.y),
                    ax=as.double(a1.x),
                    Py=as.double(P1.y),
                    Px=as.double(P1.x),
                    lonadj=as.double(lonAdj),
                    N=as.integer(N),
                    lly=as.double(0),
                    llx=as.double(0),
                    package="crawl")
    return(-2 * (out$lly + out$llx))
}
